File: Graph Theory Peaks/Union Find/Q1. Longest Consecutive Sequence/Q1_LongestConsecutiveSequence.py
class Solution(object):
    def longestConsecutive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # Convert the list to a set for fast O(1) lookups
        num_set = set(nums)
        
        # Variable to keep track of the longest consecutive sequence
        longest = 0

        # Loop through each number in the set
        for num in num_set:
            # Only start counting if num is the start of a sequence
            # (i.e., num-1 is not in the set)
            if num - 1 not in num_set:
                current_num = num        # Current number in the sequence
                current_streak = 1       # Start with a streak of length 1

                # Count consecutive numbers in the sequence
                while current_num + 1 in num_set:
                    current_num += 1      # Move to the next consecutive number
                    current_streak += 1   # Increase the current streak length

                # Update the longest streak found so far
                longest = max(longest, current_streak)

        # Return the length of the longest consecutive sequence
        return longest

File: Graph Theory Peaks/Union Find/Q1. Longest Consecutive Sequence/Q1_LongestConsecutiveSequence.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int longestConsecutive(int[] nums) {
        // Use a HashSet to store all numbers (fast lookup in O(1))
        Set<Integer> set = new HashSet<>();
        for (int num : nums) {
            set.add(num); // Add each number into the set
        }

        int longest = 0; // To keep track of the longest sequence length

        // Loop through each number in the set
        for (int num : set) {
            // Check if this number could be the start of a sequence
            // (Only start if num-1 is not in the set)
            if (!set.contains(num - 1)) {
                int currentNum = num; // Current number in the sequence
                int currentStreak = 1; // Start with length 1

                // Keep checking the next consecutive numbers
                while (set.contains(currentNum + 1)) {
                    currentNum++;       // Move to next number
                    currentStreak++;    // Increase sequence length
                }

                // Update the longest streak found so far
                longest = Math.max(longest, currentStreak);
            }
        }

        // Return the length of the longest consecutive sequence
        return longest;
    }
}
